orchestrator/lib/otel/tracing_middleware.go

```diff
@@ -0,0 +1,53 @@
+package otel
+
+import (
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
+	"go.opentelemetry.io/otel/trace"
+	"net/http"
+)
+
+func HandlerWithTracing(tracer trace.Tracer, operationName string, handler http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx, span := tracer.Start(
+			r.Context(),
+			operationName,
+			trace.WithSpanKind(trace.SpanKindServer),
+			trace.WithAttributes(
+				semconv.HTTPMethodKey.String(r.Method),
+				semconv.HTTPURLKey.String(r.URL.String()),
+				semconv.HostNameKey.String(r.Host),
+				attribute.String("tenant.id", getTenantFromRequest(r)),
+			),
+		)
+		defer span.End()
+
+		wrapped := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}
+		handler(wrapped, r.WithContext(ctx))
+
+		span.SetAttributes(semconv.HTTPStatusCodeKey.Int(wrapped.statusCode))
+		if wrapped.statusCode >= 400 {
+			span.SetStatus(codes.Error, http.StatusText(wrapped.statusCode))
+		} else {
+			span.SetStatus(codes.Ok, "")
+		}
+	}
+}
+
+type responseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (rw *responseWriter) WriteHeader(code int) {
+	rw.statusCode = code
+	rw.ResponseWriter.WriteHeader(code)
+}
+
+func getTenantFromRequest(r *http.Request) string {
+	if tenant := r.PathValue("tenant"); tenant != "" {
+		return tenant
+	}
+	return "unknown"
```

Comment:
Better to not set the attribute at all? Because some endpoints are not tenant-specific.